cupy/core/core.pyx

```diff
@@ -1320,7 +1320,9 @@ cdef class ndarray:
                 and slices == slice(None, None, None)
                 and isinstance(value, numpy.ndarray)
         ):
-            if self.dtype == value.dtype and self.shape == value.shape:
+            if (self.dtype == value.dtype
+                    and self.shape == value.shape
+                    and (self.flags.f_contiguous or self.flags.c_contiguous)):
```

Comment:
I think this will disallow copies between two (possibly non-contiguous) views that have the same strides, which is likely a regression. Looks to me that the assumption in a few lines below is wrong; it should be:
```python
if self.strides == value.strides:
    ptr = ctypes.c_void_p(value.__array_interface__['data'][0])
elif self._f_contiguous or self._c_contiguous:  # note we should use the C flags for performance
    # set up ptr and friends
else:
    raise

# do copy
```

Reply:
I don't see how you can copy into a non-contiguous memory area with a single cudaMemcpy().

Reply:
That is a good point! Then I think we should not need to check the strides below, and always require the arrays being contiguous.

Reply:
Allowing the source array to be non-contigous was an [explicit goal](https://github.com/cupy/cupy/pull/2079#issuecomment-486167028) in #2079, so I think we should continue to handle it by rearranging the source data on the CPU side before the cudaMemcpy.

Reply:
The source array can be non-contiguous as long as we do this rearrangement on CPU like you said, but the destination must be contiguous. Sorry for not being clear.

Reply:
I could eliminate the strides test, and always do a `ravel()`. This is okay since `ravel()` will not do anything if the array is already in the desired form. I don't know if this is the transform you had in mind @leofang . See https://github.com/cupy/cupy/pull/3928/commits/477785846566ab864e4d9488ce77f967bfcc6406